src/recipe_dictionary.cpp

```diff
@@ -36,15 +36,16 @@ static std::set<const recipe *> null_match;
 
 static DynamicDataLoader::deferred_json deferred;
 
-const recipe &recipe_dictionary::operator[]( const std::string &id ) const
+template<>
+const recipe &string_id<recipe>::obj() const
 {
-    auto iter = recipes.find( id );
-    return iter != recipes.end() ? iter->second : null_recipe;
+    auto iter = recipe_dict.recipes.find( *this );
+    return iter != recipe_dict.recipes.end() ? iter->second : null_recipe;
```

Comment:
Other identifiers don't quite work like this. Thy implement `string_id<T>::is_valid()` (as you're aware since you authored it) and display a `debugmsg` if the dereferenced object does not exist. Why the discrepancy here?